components/autofill/src/db/store.rs

```diff
@@ -11,50 +11,163 @@ use rusqlite::{
     Connection,
 };
 use sql_support::{self, ConnExt};
+use std::cell::RefCell;
 use std::path::Path;
-use std::sync::{Arc, Mutex};
+use std::sync::{Arc, Mutex, Weak};
 use sync15_traits::SyncEngine;
 use sync_guid::Guid;
 
-#[allow(dead_code)]
+// Our "sync manager" will use whatever is stashed here.
+lazy_static::lazy_static! {
+    // Far out:
+    // Outer mutex: just taken long enough to update the inner stuff - needed
+    //              to wrap the RefCell as they aren't `Sync`
+    // RefCell: So I can replace what it holds. Normally you'd use `get_ref()`
+    //          on the mutex, but that requires the mutex to be declared as
+    //          `mut` which is apparently impossible.
+    // [Arc/Weak]<Mutex<StoreImpl>>: What the sync manager actually needs.
+    pub static ref STORE_FOR_MANAGER: Mutex<RefCell<Weak<Mutex<StoreImpl>>>> = Mutex::new(RefCell::new(Weak::new()));
+}
+
+// This is the type that uniffi exposes. It holds an `Arc<Mutex<>>` around the
+// actual implementation, because we need access to the `Arc<Mutex<>>` for
+// integration with the sync manager. Our UDL also defines us as being
+// "ThreadSafe" to we manage our own locking via the `Mutex` - but this means
+// in practice we are actually held via an `Arc<Arc<Mutex<>>>` - but one day
+// https://github.com/mozilla/uniffi-rs/issues/417 will give us access to the
+// `Arc<>` uniffi owns, which means we can drop this mess entirely (ie, `Store`)
+// can then just have a mutex around the DB.
 pub struct Store {
```

Comment:
Does the `Store` struct need to be public? It looks like `StoreImpl` is the type that is being imported.

Reply:
This got confusing quickly and sadly they *both* need to be pub. `Store` needs to be pub because the sync manager uses it (StoreImpl has the `Arc<Store>` which we clone and hand to the sync manager, so it gets the `Store`) and `Store` needs to be pub because autofill-utils needs it to get the sync engines.

So yeah, this sucks and I added some comments, but hopefully we can get rid of this confusion with the proposed autofill enhancements.